### PR TITLE
feat: Hook up chunk grouping into provider

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/1645

# Rationale:
As part of #600 we need to combine chunks which may contain "duplicate" primary keys. You can read more background in the [design document](https://docs.google.com/document/d/1iD_Q2kpC2Kopo4DEq_Oym4jEvaz-SfUUVmr_dpMcikc/edit#)

Kudos to @NGA-TRAN for making it so clear what was needed.c

# Changes:
1. Add a `may_contain_pk_duplicates` function in PartitionChunk to signify if the chunk has potential duplicates *within* itself
2. Implement `split_overlapped_chunks` in terms of group_duplicates and may_contain_duplicates
3. tests for same

### CODE REVIEW

File: query/src/provider.rs [hunk 1]
```diff
@@ -396,22 +423,32 @@ impl<C: PartitionChunk + 'static> Deduplicater<C> {
...
-        // TODO: need to discover overlaps and split them
-        // The current behavior is just like neither overlaps nor having duplicates in its own chunk
-        //self.overlapped_chunks_set = vec![];
-        //self.in_chunk_duplicates_chunks = vec![];
-        self.no_duplicates_chunks.append(&mut chunks.to_vec());
+    fn split_overlapped_chunks(&mut self, chunks: Vec<Arc<C>>) -> Result<()> {
+        // Find all groups based on statstics
+        let groups = group_potential_duplicates(chunks).context(InternalChunkGrouping)?;
+
+        for mut group in groups {
```
Comment (alamb) [inline query/src/provider.rs:440]: here is the plumbing

File: server/src/db/chunk.rs [hunk 2]
```diff
@@ -217,6 +217,13 @@ impl PartitionChunk for DbChunk {
         self.table_name.as_ref()
     }
 
+    fn may_contain_pk_duplicates(&self) -> bool {
```
Comment (alamb) [inline server/src/db/chunk.rs:220]: This is what I understand to be the case from @pauldix  and @tustvold 's design (aka that only MUB chunks will ever have duplicates within themselves). Please let me know if that is incorrect

**Review by NGA-TRAN — APPROVED** [review]
Looks great. Thanks so much @alamb for making this available so soon. So helpful for me

File: query/src/provider.rs [hunk 3]
```diff
@@ -343,9 +361,18 @@ impl<C: PartitionChunk + 'static> Deduplicater<C> {
         schema: ArrowSchemaRef,
         chunks: Vec<Arc<C>>,
         predicate: Predicate,
-    ) -> Arc<dyn ExecutionPlan> {
-        //finding overlapped chunks and put them into the right group
-        self.split_overlapped_chunks(chunks.to_vec());
+    ) -> Result<Arc<dyn ExecutionPlan>> {
```
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/provider.rs:364]: I think it was basically inevitable given the nice structure you had going here

File: query/src/provider.rs [hunk 4]
```diff
@@ -549,3 +586,58 @@ impl<C: PartitionChunk> ChunkPruner<C> for NoOpPruner {
...
+    }
+
+    fn chunk_group_ids(groups: &[Vec<Arc<TestChunk>>]) -> Vec<String> {
+        groups
+            .iter()
+            .enumerate()
+            .map(|(idx, group)| format!("Group {}: {}", idx, chunk_ids(group)))
+            .collect()
+    }
+}
```
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/provider.rs:643]: Yeah it is kind of telling that most of this PR is test setup 🤔

File: server/src/db/chunk.rs (hunk 2 above)
Comment (NGA-TRAN) [reply server/src/db/chunk.rs:220]: This is also what I understand

**Pull request merged**